## Упражнения по библиотеке Numpy

In [1]:
import numpy as np
import random as rand
import scipy.spatial.distance as dist
import time

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [28]:
a = np.random.randint(0, 10, 10)
print(a)
a[(3 < a) & (a < 8)] *= -1
print(a)

[3 1 2 8 3 2 6 6 0 1]
[ 3  1  2 -8  3  2 -6 -6  0  1]


**2.** Заменить максимальный элемент случайного массива на 0

In [15]:
a = np.random.randint(0, 10, 10)
print(a)
a[np.argmax(a)] = 0
print(a)

[6 2 0 3 8 5 3 0 0 0]
[6 2 0 3 0 5 3 0 0 0]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [24]:
a = np.random.randint(0, 10, (2, 10))
print(a)
b = np.outer(a[0], a[1]) 
print(b)

[[5 1 2 2 9 0 9 8 8 5]
 [4 0 4 4 8 0 6 9 2 4]]
[[20  0 20 20 40  0 30 45 10 20]
 [ 4  0  4  4  8  0  6  9  2  4]
 [ 8  0  8  8 16  0 12 18  4  8]
 [ 8  0  8  8 16  0 12 18  4  8]
 [36  0 36 36 72  0 54 81 18 36]
 [ 0  0  0  0  0  0  0  0  0  0]
 [36  0 36 36 72  0 54 81 18 36]
 [32  0 32 32 64  0 48 72 16 32]
 [32  0 32 32 64  0 48 72 16 32]
 [20  0 20 20 40  0 30 45 10 20]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [144]:
a = np.random.randint(0, 5, (3, 8))
b = np.random.randint(0, 5, (2, 2))
print(a)
print(b)
for i in range(len(a)):
    if False in np.isin(b, a[i]):
        continue
    print(a[i])    

[[3 0 0 3 0 0 2 3]
 [3 1 1 3 4 3 1 2]
 [3 4 4 2 0 3 3 1]]
[[1 1]
 [4 2]]
[3 1 1 3 4 3 1 2]
[3 4 4 2 0 3 3 1]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)(fix&)

In [69]:
a = np.random.randint(0, 2, (10,3))
b = np.logical_and.reduce(a[: , 1:] == a[:,:-1], axis=1)
c = a[~b]
print(a)
print(c)

[[0 0 0]
 [1 0 1]
 [0 1 0]
 [1 0 1]
 [0 0 1]
 [0 0 0]
 [1 0 1]
 [0 0 1]
 [1 1 0]
 [0 0 1]]
[[1 0 1]
 [0 1 0]
 [1 0 1]
 [0 0 1]
 [1 0 1]
 [0 0 1]
 [1 1 0]
 [0 0 1]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [94]:
a = np.random.randint(1, 3, (20, 2))
print(a)
b = np.unique(a, axis=0)
print(b)

[[1 2]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [1 2]
 [1 2]
 [1 1]
 [2 1]
 [2 2]
 [2 2]
 [1 2]
 [2 1]
 [1 2]
 [2 1]
 [2 2]
 [2 1]
 [2 1]
 [2 1]
 [2 1]]
[[1 1]
 [1 2]
 [2 1]
 [2 2]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [140]:
a = [rand.sample(range(10), 3) for i in range(4)]
print(a)
b = 1
k = len(a) if len(a) < len(a[0]) else len(a[0])
for i in range(k):
    if a[i][i] != 0:
        b *= a[i][i]
print(b)

c = np.random.randint(0, 10, (4, 3))
print(c)
print(np.prod(np.trim_zeros(np.diagonal(c))))

[[6, 1, 5], [6, 1, 5], [2, 8, 1], [1, 8, 6]]
6
[[0 4 6]
 [0 9 4]
 [8 8 2]
 [3 7 4]]
18


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [146]:
a = rand.sample(range(10), 4)
b = rand.sample(range(10), 4)
print(a, b)
print(sorted(a) == sorted(b))

c = np.random.randint(0, 10, 4)
d = np.random.randint(0, 10, 4)
print(c, d)
print(np.array_equal(np.sort(c), np.sort(d)))

[1, 7, 0, 8] [7, 1, 4, 0]
False
[1 1 0 0] [0 1 1 0]
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [195]:
a = [rand.randint(0, 5) for i in range(20)]
print(a)
k = max([a[i] for i in range(1, len(a)) if a[i - 1] == 0])
print(k)

b = np.random.randint(0, 10, 20)
print(b)
t = np.where((b==0) & (b!=b[-1]))
print((b[t[0]-1]).max())

[0, 2, 0, 2, 4, 2, 0, 3, 5, 3, 0, 1, 3, 2, 4, 0, 4, 2, 4, 3]
4
[3 4 3 5 2 6 7 5 7 7 9 9 5 1 3 9 9 0 1 3]
9


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [4]:
a = [rand.randint(0, 5) for i in range(20)]
print(a)
vec1 = []
vec2 = []
counter = 1
for i in range(1, len(a)):
    if a[i] != a[i-1]:
        vec1.append(a[i-1])
        vec2.append(counter)
        counter = 1
        continue
    counter+=1
print((vec1, vec2))
        
    
b = np.random.randint(0, 10, 20)
print(b)
y = b[1:] != b[:-1]     
i = np.append(np.where(y), len(b) - 1)
z = np.diff(np.append(-1, i))  
print(z, b[i])    

[2, 4, 2, 5, 0, 3, 3, 4, 1, 3, 5, 2, 0, 0, 3, 4, 3, 5, 1, 5]
([2, 4, 2, 5, 0, 3, 4, 1, 3, 5, 2, 0, 3, 4, 3, 5, 1], [1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1])
[  1 666   1   1   1   3   3   3  23  23   1   8  42  42 555 555 555   1
 777]
[1 1 3 3 2 1 1 2 3 1 1] [  1 666   1   3  23   1   8  42 555   1 777]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [2]:
e = [[5, 4, 0], [3, 3, 3], [3, 5, 6], [1, 3, 4]]
c = [[2, 4, 0], [3, 3, 3], [3, 5, 6], [1, 3, 4]]

start_time = time.time()
distance = []
for x in e:
    row = []
    for y in c:
        row.append(pow((sum([(a - b) ** 2 for a, b in zip(x, y)])), 1/2))
    distance.append(row)
print(distance)
print("Python: %s seconds" % (time.time() - start_time))

a = np.array(e)
b = np.array(c)
start_time = time.time()
p = np.add.outer(np.sum(a**2, axis=1), np.sum(b**2, axis=1))
n = np.dot(a, b.T)
print(np.sqrt(p - 2*n))
print("NumPy: %s seconds" % (time.time() - start_time))

start_time = time.time()
print(dist.cdist(e, c))
print("SciPy: %s seconds" % (time.time() - start_time))

[[3.0, 3.7416573867739413, 6.4031242374328485, 5.744562646538029], [3.3166247903554, 0.0, 3.605551275463989, 2.23606797749979], [6.164414002968976, 3.605551275463989, 0.0, 3.4641016151377544], [4.242640687119285, 2.23606797749979, 3.4641016151377544, 0.0]]
Python: 0.00046825408935546875 seconds
[[3.         3.74165739 6.40312424 5.74456265]
 [3.31662479 0.         3.60555128 2.23606798]
 [6.164414   3.60555128 0.         3.46410162]
 [4.24264069 2.23606798 3.46410162 0.        ]]
NumPy: 0.04676556587219238 seconds
[[3.         3.74165739 6.40312424 5.74456265]
 [3.31662479 0.         3.60555128 2.23606798]
 [6.164414   3.60555128 0.         3.46410162]
 [4.24264069 2.23606798 3.46410162 0.        ]]
SciPy: 0.0007905960083007812 seconds


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [129]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [160]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [133]:
average_calories = np.mean(calorie_stats)
average_calories 

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [132]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [134]:
median_calories = np.median(calorie_stats)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [157]:
print(np.percentile(calorie_stats, 60))
print(np.percentile(calorie_stats, 30))
print(np.percentile(calorie_stats, 20))
print(np.percentile(calorie_stats, 10))
print(np.percentile(calorie_stats, 3.3))
print(np.percentile(calorie_stats, 2))
nth_percentile = np.percentile(calorie_stats, 3.3)

110.0
100.0
100.0
90.0
60.16
50.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [178]:
more_calories = '{:.1%}'.format(np.mean((60 < calorie_stats)))
more_calories 

'96.1%'

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [180]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Исходя из подсчётов выше, можно сказать, что большинство конкурентов наших хлопьев имеет повышенную калорийность. Этим можно воспользоваться для того, чтобы пропиарить наши хлопья среди потенциальных покупателей. Хотя и есть хлопья, в которых калорий меньше, мы всё равно входим в 96.1% процент лучших хлопьев по этому показателю. Также стоит сказать, что разброс в 19.357 относительно среднего значения 110 - достаточно небольшой. то значит, что конкуренты не особо заботятся о своих покупателях и не пытаются сделать что-либо действительн новое и полезное.